server/pulp/server/controllers/repository.py

```diff
@@ -789,16 +780,22 @@ def sync(repo_id, sync_config_override=None, scheduled_call_id=None):
         else:
             msg = _('Plugin type [%s] on repo [%s] did not return a valid sync report')
             _logger.warn(msg % (repo_importer['importer_type_id'], repo_obj.repo_id))
-            added_count = updated_count = removed_count = -1  # None?
             summary = details = msg
             result_code = RepoSyncResult.RESULT_ERROR  # RESULT_UNKNOWN?
 
-        sync_result = RepoSyncResult.expected_result(
-            repo_obj.repo_id, repo_importer['id'], repo_importer['importer_type_id'],
-            sync_start_timestamp, sync_end_timestamp, added_count, updated_count, removed_count,
-            summary, details, result_code)
+        sync_result, sync_end_timestamp = _reposync_result(repo_obj, repo_importer,
+                                                           sync_start_timestamp, summary, details,
+                                                           result_code, before_sync_unit_count)
 
     finally:
+        if sync_result is None:
```

Comment:
The main problem with the current design is that sys.exit() causes the celery worker to exit. We don't want that we just want the task to stop immediately. To resolve this, instead of doing this here, let's do this instead:

1) in async/tasks.py define an exception that will be raised. Call it `CancelException`
2) Raise `CancelException` where sys.exit() is today
3) Write a plugin API release note about this change
4) Make a decorator called `@cancelable` which effectively does:

```
try:
    the_decorated_func()
except CancelException:
    # log that the task is being canceled
```

5) decorate sync and publish tasks in platform with the `@cancelable` decorator